exchange/index.go

```diff
@@ -453,6 +465,84 @@ func (idx *Index) evict(size uint64) uint64 {
 	return evicted
 }
 
+func (idx *Index) garbageCollect(ref *DataRef) error {
+	cs := cid.NewSet()
+	link := cidlink.Link{Cid: ref.PayloadCID}
+	chooser := dagpb.AddDagPBSupportToChooser(func(ipld.Link, ipld.LinkContext) (ipld.NodePrototype, error) {
+		return basicnode.Prototype.Any, nil
+	})
+
+	// The root node could be a raw node so we need to select the builder accordingly
+	nodeType, err := chooser(link, ipld.LinkContext{})
+	if err != nil {
+		return err
+	}
+	builder := nodeType.NewBuilder()
+
+	// We make a custom loader to intercept when each block is read during the traversal
+	makeLoader := func(bs blockstore.GCBlockstore) ipld.Loader {
+		return func(lnk ipld.Link, lnkCtx ipld.LinkContext) (io.Reader, error) {
+			c, ok := lnk.(cidlink.Link)
+			if !ok {
+				return nil, fmt.Errorf("incorrect Link Type")
+			}
+
+			block, err := bs.Get(c.Cid)
+			if err != nil {
+				return nil, err
+			}
+
+			reader := bytes.NewReader(block.RawData())
+
+			v := new(DataRef)
+			if err := v.UnmarshalCBOR(reader); err != nil {
+				return reader, err
+			}
+
+			cs.Add(v.PayloadCID)
```

Comment:
@tchardin Why can't I run `idx.bstore.DeleteBlock(c)`insted ?

Reply:
The way I thought about it, the eviction method might evict multiple refs, so you want the eviction method simply to add CIDs to a set that has its own lock on the index struct. Once all the root CIDs have been marked the operation can return. Then we expose a separate public method like `Index.GC()` which we can trigger and that will take any CIDs in the set and delete them. What do you think?

Reply:
Ok I think I got it, but then when do we trigger the index.GC() ?

Reply:
since the eviction is only called in the setRef, we would run index.GC() right after?

Reply:
I think we want garbage collection to be decoupled from SetRef, so we have more flexibility to try out different strategies. Basically we could start by calling it in the `node.Commit` call and if that consumes too much resources we could call it periodically so that we can call Commit a bunch of times and garbage collect every once in a while. We need to measure the strain on the system

Reply:
Ok sounds good like that 👍

Reply:
Why not doing both actually? Like we call GC every time in the Commit, but if the last one was made not so long ago, we don't do it (we can introduce a lastGC variable that contains the time the last GC was made).
On top of that, we can have a loop with a ticker, that checks once in a while if there are things to be GC